File: android/app/src/main/java/com/nvp/node/net/ApiClient.kt
package com.nvp.node.net

import com.nvp.node.Config
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.MediaType.Companion.toMediaType
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class Job(val jobId: String, val model: String, val prompt: String, val maxTokens: Int, val reasoning: Boolean)
data class SubmitResult(val accepted: Boolean, val credited: Double, val balance: Double, val reason: String?)

/** Thin HTTP client to the coordinator (register / heartbeat / nextJob / submit). */
class ApiClient {
    private val json = "application/json".toMediaType()
    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(35, TimeUnit.SECONDS) // long-poll
        .build()

    private fun base() = Config.coordinatorUrl.trimEnd('/')
    private fun authed(b: Request.Builder): Request.Builder {
        Config.apiKey?.let { b.header("Authorization", "Bearer $it") }
        return b
    }

    /** Register the worker; returns the issued device API key (if any). */
    fun register(devicePubkey: String): String? {
        val body = JSONObject()
            .put("device_pubkey", devicePubkey)
            .put("platform", "android")
            .put("model_caps", JSONArray(Config.modelCaps))
            .toString().toRequestBody(json)
        val req = Request.Builder().url(base() + "/api/workers/register").post(body).build()
        client.newCall(req).execute().use { r ->
            if (!r.isSuccessful) return null
            val o = JSONObject(r.body?.string() ?: "{}")
            return o.optString("api_key", null) ?: o.optString("apiKey", null)
        }
    }

    fun heartbeat() {
        runCatching {
            val req = authed(Request.Builder().url(base() + "/api/me/worker")).build()
            client.newCall(req).execute().use { }
        }
    }

    fun health(): Boolean = runCatching {
        val req = Request.Builder().url(base() + "/api/health").build()
        client.newCall(req).execute().use { it.isSuccessful }
    }.getOrDefault(false)

    /** Long-poll for the next job; null on 204/no job. */
    fun nextJob(models: List<String>): Job? {
        val url = base() + "/api/jobs/next?models=" + models.joinToString(",")
        val req = authed(Request.Builder().url(url)).build()
        client.newCall(req).execute().use { r ->
            if (r.code == 204 || !r.isSuccessful) return null
            val o = JSONObject(r.body?.string() ?: "{}")
            val p = o.optJSONObject("params") ?: JSONObject()
            return Job(
                jobId = o.optString("job_id"),
                model = o.optString("model"),
                prompt = o.optString("prompt"),
                maxTokens = p.optInt("max_tokens", Config.defaultMaxTokens),
                reasoning = p.optBoolean("reasoning", false),
            )
        }
    }

    fun submitResult(jobId: String, output: String, latencyMs: Int, tokensOut: Int): SubmitResult? {
        val body = JSONObject()
            .put("output", output).put("latency_ms", latencyMs).put("tokens_out", tokensOut)
            .toString().toRequestBody(json)
        val req = authed(Request.Builder().url(base() + "/api/jobs/$jobId/result")).post(body).build()
        client.newCall(req).execute().use { r ->
            if (!r.isSuccessful) return null
            val o = JSONObject(r.body?.string() ?: "{}")
            return SubmitResult(o.optBoolean("accepted"), o.optDouble("credited", 0.0), o.optDouble("balance", 0.0), o.optString("reason", null))
        }
    }
}

File: macos/NVPMac/APIClient.swift
import Foundation

struct Job { let jobId: String; let model: String; let prompt: String; let maxTokens: Int }
struct SubmitResult { let accepted: Bool; let credited: Double; let balance: Double }

/// HTTP client to the coordinator (register / heartbeat / nextJob / submit).
final class APIClient {
    private let session: URLSession = {
        let c = URLSessionConfiguration.default
        c.timeoutIntervalForRequest = 35 // long-poll
        return URLSession(configuration: c)
    }()

    private func base() -> String { Config.coordinatorURL.hasSuffix("/") ? String(Config.coordinatorURL.dropLast()) : Config.coordinatorURL }
    private func req(_ path: String, method: String = "GET", body: [String: Any]? = nil) -> URLRequest? {
        guard let url = URL(string: base() + path) else { return nil }
        var r = URLRequest(url: url); r.httpMethod = method
        if let key = Config.apiKey { r.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization") }
        if let body { r.setValue("application/json", forHTTPHeaderField: "Content-Type"); r.httpBody = try? JSONSerialization.data(withJSONObject: body) }
        return r
    }

    func health() async -> Bool {
        guard let r = req("/api/health") else { return false }
        guard let (_, resp) = try? await session.data(for: r), let h = resp as? HTTPURLResponse else { return false }
        return (200..<300).contains(h.statusCode)
    }

    func register() async -> String? {
        guard let r = req("/api/workers/register", method: "POST", body: [
            "device_pubkey": Config.deviceId, "platform": "macos", "model_caps": Config.modelCaps,
        ]) else { return nil }
        guard let (d, resp) = try? await session.data(for: r), let h = resp as? HTTPURLResponse, (200..<300).contains(h.statusCode) else { return nil }
        let o = (try? JSONSerialization.jsonObject(with: d)) as? [String: Any]
        return (o?["api_key"] as? String) ?? (o?["apiKey"] as? String)
    }

    func heartbeat() async { if let r = req("/api/me/worker") { _ = try? await session.data(for: r) } }

    func nextJob() async -> Job? {
        guard let r = req("/api/jobs/next?models=" + Config.modelCaps.joined(separator: ",")) else { return nil }
        guard let (d, resp) = try? await session.data(for: r), let h = resp as? HTTPURLResponse else { return nil }
        if h.statusCode == 204 || !(200..<300).contains(h.statusCode) { return nil }
        guard let o = (try? JSONSerialization.jsonObject(with: d)) as? [String: Any] else { return nil }
        let p = o["params"] as? [String: Any]
        return Job(jobId: o["job_id"] as? String ?? "", model: o["model"] as? String ?? "",
                   prompt: o["prompt"] as? String ?? "", maxTokens: (p?["max_tokens"] as? Int) ?? Config.maxTokens)
    }

    func submit(jobId: String, output: String, latencyMs: Int, tokensOut: Int) async -> SubmitResult? {
        guard let r = req("/api/jobs/\(jobId)/result", method: "POST", body: [
            "output": output, "latency_ms": latencyMs, "tokens_out": tokensOut,
        ]) else { return nil }
        guard let (d, resp) = try? await session.data(for: r), let h = resp as? HTTPURLResponse, (200..<300).contains(h.statusCode) else { return nil }
        let o = (try? JSONSerialization.jsonObject(with: d)) as? [String: Any] ?? [:]
        return SubmitResult(accepted: o["accepted"] as? Bool ?? false, credited: o["credited"] as? Double ?? 0, balance: o["balance"] as? Double ?? 0)
    }
}
